CMakeLists.txt

```diff
@@ -431,23 +431,22 @@ set(IREE_CUDA_LIBDEVICE_PATH "" CACHE STRING "Absolute path to an appropriate li
 # If any CUDA features are being built, try to locate a CUDA SDK. We will fall
 # back to this as needed for specific features.
 if(IREE_TARGET_BACKEND_CUDA OR IREE_HAL_DRIVER_CUDA)
-  find_package(CUDAToolkit)
-
-  if(NOT CUDAToolkit_FOUND)
-    if(DEFINED ENV{IREE_CUDA_DEPS_DIR})
-      # We define the magic IREE_CUDA_DEPS_DIR env var in our CI docker images if we
-      # have a stripped down CUDA toolkit suitable for compiling available. We
-      # trigger on this below as a fallback for locating headers and libdevice
-      # files.
-      set(CUDAToolkit_ROOT "$ENV{IREE_CUDA_DEPS_DIR}")
-      message(STATUS "CUDA SDK not found by CMake but using IREE_CUDA_DEPS = ${CUDAToolkit_ROOT}")
-    else()
-      # If we haven't found CUDA deps, download at least enough to build for CUDA.
-      # This will define IREE_CUDA_DOWNLOAD_LIBDEVICE_PATH & IREE_CUDA_DOWNLOAD_INCLUDE_PATH
-      # vars with the target deps.
-      message(STATUS "CUDA SDK not found by CMake but downloading dependencies")
+  if(DEFINED ENV{IREE_CUDA_DEPS_DIR})
+    # We define the magic IREE_CUDA_DEPS_DIR env var in our CI docker images if we
+    # have a stripped down CUDA toolkit suitable for compiling available. We
+    # trigger on this below as a fallback for locating headers and libdevice
+    # files. See build_tools/docker/base/fetch_cuda_deps.sh for what this
+    # includes (it does not include enough for the normal CMake toolkit search
+    # to succeed).
+    set(CUDAToolkit_ROOT "$ENV{IREE_CUDA_DEPS_DIR}")
+    message(STATUS "Using CUDA minimal deps for CI using IREE_CUDA_DEPS = ${CUDAToolkit_ROOT}")
+  else()
+    if(NOT CUDAToolkit_ROOT)
+      # Auto-download and setup CUDAToolkit_ROOT
+      message(STATUS "CUDAToolkit_ROOT not defined: downloading expected version")
       add_subdirectory(build_tools/third_party/cuda EXCLUDE_FROM_ALL)
     endif()
+    find_package(CUDAToolkit REQUIRED)
```

Comment:
Just want to double check the intention of moving the ` find_package(CUDAToolkit REQUIRED)` here? I don't think this detects the packages downloaded with `add_subdirectory(build_tools/third_party/cuda EXCLUDE_FROM_ALL)` since those aren't on any system path (they provide `IREE_CUDA_DOWNLOAD_LIBDEVICE_PATH & IREE_CUDA_DOWNLOAD_INCLUDE_PATH` for their location). 

Also if a CUDA toolkit is found I'd expect there is no need to download more unless there are some additional dependencies missing?

Reply:
I have found that doing a find_package(CUDAToolkit) is unsafe to do more than once, so I restructured it to avoid what I was seeing as a lot of variability: either the user specifies a CUDAToolkit_ROOT and that is honored, or they don't and something suitable is downloaded for them. Since the above add_subdirectory sets the CUDAToolkit_ROOT, it is picked up here.